quipucords/api/scanjob/queryset.py

```diff
@@ -0,0 +1,68 @@
+"""Module for ScanTaskQuerySet."""
+
+from django.db.models import Case, F, Q, QuerySet, Sum, When
+
+from api.scantask.model import ScanTask
+
+
+class ScanJobQuerySet(QuerySet):
+    """Custom QuerySet for ScanJob."""
+
+    def with_counts(self):
+        """Annotate ScanJob with system counts from associated tasks."""
+        return self.annotate(
+            systems_count=Sum(
+                "tasks__systems_count",
+                filter=Q(tasks__scan_type=ScanTask.SCAN_TYPE_CONNECT),
+                default=0,
+            ),
+            systems_scanned=Sum(
+                "tasks__systems_scanned",
+                # only counts of tasks with the same scan_type should be considered
+                filter=Q(tasks__scan_type=F("scan_type")),
+                default=0,
+            ),
+            # case required for following annotations because inpection type scansjobs
+            # sum failed/unreachable values from both connect and inspect tasks, while
+            # connection type scanjobs only look for connection tasks
```

Comment:
Although this sounds reasonable, in practice, do we ever _actually_ have unexpected scan tasks associated with a scan job? For example, do we ever create a connection-type scanjob that has an inspection-type scan type, and if we did, would it ever have non-zero count for systems failed? I'm just wondering if there are specific situations that prevent us from simplifying the `Sum` without any `tasks__scan_type` filter.

Full disclosure: I have not gone through and reviewed the scan job/task code recently. I'm just working out what makes logic sense in my head.

Reply:
I *believe* connection type scanjobs only have a single scan task.

It occurred to me a simple sum all, without the case, could be enough - specially because in production fingerprint type tasks always have the value zero in production. But then I started wondering we might want to change that so counts in fingerprint type task have a meaning.

Reply:
My instinct says KISS, but if you believe the counts may change in the near future, I'll defer to your judgment. 🙂

Reply:
KISS is always a good point. IF we need the complexity later, we defer to add it later.
I'll first confirm if a connection type scan only has a connection task first.

the side effect would be loss of compatibility with the legacy method that does counts, but it doesn't really matter on realistic scenarios.

Reply:
Took some time researching and experimenting. Connect type jobs only have connect type tasks, so we could get rid of the case and both connect/inspect type scans would work as intended.

However, for fingerprint type scans things would be a bit odd because we would change the behavior from showing meaningless zero values (which I'd like to change soon ™️ ) to now meaningless counts from other tasks.

Reply:
query in postgresql

```
SELECT
    "api_scanjob"."id",
    "api_scanjob"."scan_type",
    "api_scanjob"."status",
    "api_scanjob"."status_message",
    "api_scanjob"."options_id",
    "api_scanjob"."start_time",
    "api_scanjob"."end_time",
    "api_scanjob"."scan_id",
    "api_scanjob"."connection_results_id",
    "api_scanjob"."inspection_results_id",
    "api_scanjob"."report_id",
    CASE
        WHEN "api_scanjob"."scan_type" = inspect THEN COALESCE(
            SUM("api_scantask"."systems_count") FILTER (
                WHERE
                    "api_scantask"."scan_type" = connect
            ),
            0
        )
        ELSE COALESCE(
            SUM("api_scantask"."systems_count") FILTER (
                WHERE
                    "api_scantask"."scan_type" = ("api_scanjob"."scan_type")
            ),
            0
        )
    END AS "systems_count",
    COALESCE(
        SUM("api_scantask"."systems_scanned") FILTER (
            WHERE
                "api_scantask"."scan_type" = ("api_scanjob"."scan_type")
        ),
        0
    ) AS "systems_scanned",
    CASE
        WHEN "api_scanjob"."scan_type" = inspect THEN COALESCE(
            SUM("api_scantask"."systems_failed") FILTER (
                WHERE
                    "api_scantask"."scan_type" IN (connect, inspect)
            ),
            0
        )
        ELSE COALESCE(
            SUM("api_scantask"."systems_failed") FILTER (
                WHERE
                    "api_scantask"."scan_type" = ("api_scanjob"."scan_type")
            ),
            0
        )
    END AS "systems_failed",
    CASE
        WHEN "api_scanjob"."scan_type" = inspect THEN COALESCE(
            SUM("api_scantask"."systems_unreachable") FILTER (
                WHERE
                    "api_scantask"."scan_type" IN (connect, inspect)
            ),
            0
        )
        ELSE COALESCE(
            SUM("api_scantask"."systems_unreachable") FILTER (
                WHERE
                    "api_scantask"."scan_type" = ("api_scanjob"."scan_type")
            ),
            0
        )
    END AS "systems_unreachable"
FROM
    "api_scanjob"
    LEFT OUTER JOIN "api_scantask" ON ("api_scanjob"."id" = "api_scantask"."job_id")
GROUP BY
    "api_scanjob"."id"
```